### PR TITLE
Introduce caching profiles

### PR DESCRIPTION
**Pull request by mollyIV**
### Short description 📝

This pull request is introducing caching profiles. Initial work has been done in https://github.com/tuist/tuist/pull/2221. 

The caching profile allows users to specify how targets will be cached, for example setting a configuration. Instead of passing the configuration through CLI arguments, developers can define a profile and reference it with `tuist cache warm --profile MyProfile`. You can define a caching profile in `Config.swift`, for example:

```
let config = Config(cache: .cache(profiles: [.profile(name: "Simulator", configuration: "Debug")]))
```

The following commands has been affected: 

* `tuist cache warm --profile Simulator`
* `tuist focus TargetA --profile Simulator`
* `tuist cache print-hashes --profile Simulator`

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] A developer other than the author has verified that the changes work as expected.
- [x] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Sources/TuistCache/ContentHashing/CacheProfileContentHasher.swift
Comment (fortmarek) [inline Sources/TuistCache/ContentHashing/CacheProfileContentHasher.swift:23]: might be worth writing to the internal documentation of `Cache.Profile` a warning that if one adds a property, it should be added here, too. I also thought about how we could check this statically but only thing that could probably achieve this is using `Mirror`, so I think some written warning should do

File: Sources/TuistKit/Commands/Cache/CachePrintHashesCommand.swift
Comment (fortmarek) [inline Sources/TuistKit/Commands/Cache/CachePrintHashesCommand.swift:23]: `CachePrintHashesCommand` has the same options as `CacheWarmCommand` - rather than duplicating those, it might be useful to define [OptionGroup](https://github.com/apple/swift-argument-parser/blob/main/Examples/math/main.swift#L57) called `CacheOptions`?

File: Sources/TuistKit/Services/Cache/CacheProfileResolver.swift
Comment (fortmarek) [inline Sources/TuistKit/Services/Cache/CacheProfileResolver.swift:6]: to stay consistent with the rest of the codebase, I'd name this explicitly `CacheProfileResolverError`

File: Tests/TuistKitTests/Services/Cache/CachePrintHashesServiceTests.swift
Comment (fortmarek) [inline Tests/TuistKitTests/Services/Cache/CachePrintHashesServiceTests.swift:113]: nitpick: This should probably be in `Given` section and `When` should be used before `subject.run(...)`

**Review by fortmarek — COMMENTED** [review]
Well done - I have only smaller things to add but the work seems otherwise solid to me 🙌 although, some cache expert should look into this, too 🙁

File: Sources/TuistCache/ContentHashing/CacheProfileContentHasher.swift
Comment (mollyIV) [reply Sources/TuistCache/ContentHashing/CacheProfileContentHasher.swift:23]: Sure, good idea, I added a warning message to the `Cache.Profile` struct 👍

File: Sources/TuistKit/Commands/Cache/CachePrintHashesCommand.swift
Comment (mollyIV) [reply Sources/TuistKit/Commands/Cache/CachePrintHashesCommand.swift]: I did not know about an OptionGroup, that's great! Applied the changes.

**Comment by mollyIV** [comment]
Hey @fortmarek, 

thank you for the review. I applied the changes, cheers 😊

File: Sources/TuistKit/Commands/Cache/CacheWarmCommand.swift
Comment (pepibumur) [inline Sources/TuistKit/Commands/Cache/CacheWarmCommand.swift:23]: Nitpick: What about having `CacheOptions` in its own file `CacheOptions.swift`?

**Review by pepibumur — APPROVED** [review]
It looks great. Great job @mollyIV 👏

**Review by fortmarek — APPROVED** [review]
LGTM once the merge conflicts are solved

File: Sources/TuistKit/Commands/Cache/CacheWarmCommand.swift
Comment (mollyIV) [reply Sources/TuistKit/Commands/Cache/CacheWarmCommand.swift]: The caching options has been extracted to a separate file 👍

**Comment by mollyIV** [comment]
cache-frameworks acceptance tests are failing on `main` branch: https://github.com/tuist/tuist/runs/1872121367.

**Comment by fortmarek** [comment]
Seems to be [fixed](https://github.com/tuist/tuist/runs/1873587407)

Please rebase @mollyIV, sorry for the back and forth 😞

**Comment by mollyIV** [comment]
Hey @fortmarek 👋 

No worries, I did a rebase and the checks are green 🟢

**Comment by fortmarek** [comment]
Well done 👏  merged!

**Pull request merged**
